nilearn/reporting/glm_reporter.py

```diff
@@ -967,3 +973,289 @@ def _dataframe_to_html(df, precision, **kwargs):
     html_table = html_table.replace('border="1" ', '')
     return html_table
+
+
+def _clean_contrast_name(contrast_name):
+    """Remove non-alphanumeric characters from a contrast name.
+
+    BIDS filenames, in which the contrast name will appear as a
+    contrast-<name> key/value pair, must be alphanumeric strings.
+
+    Parameters
+    ----------
+    contrast_name : :obj:`str`
+        Contrast name to clean.
+
+    Returns
+    -------
+    new_name : :obj:`str`
+        Contrast name with non-alphanumeric characters removed.
+    """
+    new_name = ''.join(ch for ch in contrast_name if ch.isalnum())
+    if new_name != contrast_name:
+        warnings.warn(
+            'Contrast name "{}" changed to "{}"'.format(
+                contrast_name,
+                new_name
+            )
+        )
+    return new_name
+
+
+def save_glm_results(model, contrasts, out_dir='.', prefix=None):
+    """Save GLM results to BIDS-like files.
+
+    Parameters
+    ----------
+    model : :obj:`nilearn.glm.first_level.FirstLevelModel` or
+            :obj:`nilearn.glm.second_level.SecondLevelModel`
+        First- or second-level model from which to save outputs.
+    contrasts : :obj:`dict` of :obj:`str`-:obj:`numpy.ndarray` pairs
+        A dictionary containing contrasts. Keys are contrast names,
+        while values are arrays containing contrast weights.
+        The arrays may be 1D (for t-contrasts) or 2D (for F-contrasts).
+    out_dir : :obj:`str`, optional
+        Output directory for files. Default is current working directory.
+    prefix : :obj:`str` or None, optional
+        String to prepend to generated filenames.
+        If a string is provided, "_" will be added to the end.
+        Default is None.
+
+    Warning
+    -------
+    The files generated by this function are a best approximation of
+    appropriate for a GLM-based BIDS derivative.
+    However, BIDS does not currently have GLM-based derivatives supported in
+    the specification, and there is no guarantee that the files created by
+    this function will be BIDS-compatible if and when the specification
+    supports model derivatives.
+
+    Notes
+    -----
+    To output:
+    - design matrix
+    - contrast plot
+    - contrast-level stat maps
+    - model-level stat maps
+    - metadata
+    """
+    if isinstance(prefix, str) and not prefix.endswith('_'):
+        prefix += '_'
+    else:
+        prefix = ''
+
+    out_dir = os.path.abspath(out_dir)
+
+    model_level = (
+        1 if isinstance(model, glm.first_level.FirstLevelModel) else 2
+    )
+
+    # Write out design matrices to files.
+    if hasattr(model, 'design_matrices_'):
+        design_matrices = model.design_matrices_
+    else:
+        design_matrices = [model.design_matrix_]
+
+    # TODO: Assuming that cases of multiple design matrices correspond to
+    # different runs. Not sure if this is correct. Need to check.
+    for i_run, design_matrix in enumerate(design_matrices):
```

Comment:
My assumption here is that multiple design_matrices in a model will _always_ mean applying the same model to multiple runs. Is that accurate?